feat :: add soft-hyphen for multi-cell

oleksii-shyman opened the pull request:
Close #291

  - refactor existing codebase for readability
  - introduce automatic line break at soft-hyphen location

**Checklist**:

<!-- To check an item, place an "x" in the box like so: "- [x] Item description"
     Add "N/A" to the end of each line that's irrelevant to your changes -->

- [x] The GitHub pipeline is OK (green), <!-- The maintainers will trigger it if this is your 1st contribution -->
      meaning that both `pylint` (static code analyzer) and `black` (code formatter) are happy with the changes of this PR.

- [x] A unit test is covering the code added / modified by this PR

- [x] This PR is ready to be merged <!-- In your opinion, can this be merged as soon as it's reviewed? Else, this can be turned into a Draft PR -->

- [x] In case of a new feature, docstrings have been added, with also some documentation in the `docs/` folder

- [ ] A mention of the change is present in `CHANGELOG.md`

- [ ] The PR description or comment contains a picture of a cute animal (not mandatory but encouraged 😉)

<!-- Feel free to add additional comments, and to ask questions if some of those guidelines are unclear to you! -->

<!--
Once a PR is merged, maintainers will add your name to the contributors table in README.md.
If they forget, or you do not wish this to happen, please mention it.
-->

oleksii-shyman commented:
At this point it is in `draft` state.
Logics _is expected to be valid_ , but I wasn't able to run / test my code yet

oleksii-shyman commented:
basic test is working now. able to print 20+ pages of lorem ipsum using refactored approach.
output _looks_ the same to the master version, but tests are failing. identifying the root cause

gmischler commented:
That looks promising, very cool!

>  output _looks_ the same to the master version, but tests are failing. identifying the root cause

Tests breaking on a code change are often caused by non-semantic differences in the PostScript code of the PDF (sometimes simply changing insignificant whitespace).
If you look in the temporary directory where the test output is placed, then you'll find versions of both old and new PDF with the PostScript sections decompressed, which you can compare in a text editor.

The previous code uses `self._out("0 Tw")` several times to reset the word spacing, while yours doesn't. That would probably account for a technical difference in the output. If you find that this doesn't change the appearance of the output, then everything is fine, and you can replace the PDF files in the repository with your new ones (found in the same temp dir) to make the tests pass. Of course, if you should find any visible differences, you'd have to fix those first.

Lucas-C commented:
Thank you for your contribution @oleksii-shyman!

I second @gmischler answer.
You'll also find more information on tests in the documentation:
https://pyfpdf.github.io/fpdf2/Development.html#why-is-a-test-failing

oleksii-shyman commented:
I was able to fix most of the cases, only 1 pdf updated, visually validated that it has no differences with the original one.

Currently, only 2 markdown related cases are failing:
```
FAILED test/cells/test_multi_cell_markdown.py::test_multi_cell_markdown
FAILED test/cells/test_multi_cell_markdown.py::test_multi_cell_markdown_with_ttf_fonts
```
I was able to investigate the root cause, and understand why does it fail.
I originally designed this solution to consist of 2 stages:
1. Break normalized string into feasible chunks
2. Wrap feasible chunk with `self.cell` calls

However, this approach doesn't work on markdown strings, because markdown special characters (`*`, `_`) can affect the style of the font. And the style of the font can affect the width of any character.

I want to discuss how to properly solve this issue. One easy way to do it - to give up on the idea of 2 stages, just issue a call to `self.cell` as soon as feasible substring is found. In other words - keep the existing approach.
However, there are 2 problems with the existing approach:
1. underscore(`_`) and asterisk (`*`) characters width is included into the total width of the cell here -> https://github.com/oleksii-shyman/fpdf2/blob/master/fpdf/fpdf.py#L2341-L2344 . But these characters are not displayed to the user, so they should be ignored.
2. splitting the markdown string without __immediately__ taking into account the style change leads to improper width calculation. In the existing approach style is changed only after the call to `self.cell` method.

Based on the problems with the existing approach i propose another way to do it:
1. Parse normalized string into markdown chunks with `self._preload_font_styles(normalized_string, markdown)`. This function returns chunks of text with associated style. Also underscores and asterisks are removed -> `(('Lorem ipsum dolor amet, ', '', False), ('consectetur adipiscing', 'B', False), (' elit, sed do eiusmod ', '', False), ('tempor incididunt', 'I', False), (' ut labore et dolore magna aliqua.', '', False))`
2. Break the string into feasible chunks (taking into the accont the style)
3. Wrap feasible chunk with `self.cell` call. 

The problem here is that `self.cell` accepts raw text (unparsed markdown) instead of chunks of text with associated style. I plan to split `self.cell` to 2 methods
1. `self._cell` - method that accepts parsed markdown as the argument.
2. `self.cell` - with exactly the same interface as before, no caller will spot the difference. This method will parse the markdown and pass it to `self._cell` method


@gmischler @Lucas-C What do you think?

oleksii-shyman commented:
![image](https://user-images.githubusercontent.com/8827452/147865979-dcd62e6e-c305-4ede-86c8-0c910ac81870.png)
![image](https://user-images.githubusercontent.com/8827452/147866002-463fda71-bb9a-48a1-8f23-a41cebec349b.png)

Examples (ttf \ no ttf) of basic test case with soft hyphen
```
text = (  # Some text where styling occur over line breaks:
    "Lorem ipsum dolor amet, con\u00adsec\u00adte\u00adtur adipiscing elit,"
    " sed do eiusmod tempor incididunt ut labore et dolore magna aliqua."
) 
```
As you can see - there are 3 soft hyphens in the same word, but only one is present.

Lucas-C commented:
Your latest  demo-screenshot looks very promising @oleksii-shyman !
And good job really on the thorough code & options analysis!

I introduced `_preload_font_styles()` myself a few months ago.
I understand the need here to extract a sub-method from `FPDF.cell()`, that could be called by `multi_cell()`.
I think a better naming for it could be `_render_styled_cell_text()`.
I guess this method would basically contains all the code [in the `if txt` block at line 1963](https://github.com/PyFPDF/fpdf2/blob/2.4.6/fpdf/fpdf.py#L1963).
AFAIC, you can proceed down this road 😉 And thanks for your time and efforts!

@all-contributors please add @oleksii-shyman for code, design, ideas

gmischler commented:
Yeah, I expected that to be a non-trivial task, but you're getting there!

> The problem here is that `self.cell` accepts raw text (unparsed markdown) instead of chunks of text with associated style. I plan to split `self.cell` to 2 methods
> 
>     1. `self._cell` - method that accepts parsed markdown as the argument.
> 
>     2. `self.cell` - with exactly the same interface as before, no caller will spot the difference. This method will parse the markdown and pass it to `self._cell` method


Now allow me to make things even more complicated... 😁 

You may have noticed that `.write()` does something very similar to `.multicell()`, mostly just with different horizontal boundaries and different wrapping rules. It currently has its own version of the line wrapping code, most of which is a redundant copy of `.multicell()`. Eventually, the two should really use the same code, for simplicity and maintainability. No, I'm not suggesting you should do this now, that's a seperate task.
But...

When chosing your approach for `.multicell()`, you might want to check your solution does not unnecessarily stand in the way of doing that later. The approach of `.write()` is different, in that it doesn't give you all the parts in one go as with the markup option, but it lets the user feed in differently formatted chunks of text one after the other.

I haven't looked at it very closely, but I expect that both `.multicell()` and `.write()` would eventually use a common internal method to actually do the wrapping.

Given that you need to do the markdown splitting at a very early stage, wrapping `.cell()` inside `._cell()` (or `._cell_row()`/ `._render_styled_cell_text()`) seems to make a lot of sense. It can also decide which of its `cell()`s need border lines on which sides.

As an organisational detail, if you create new classes to do the formatting, then it might be a good idea to place those in a seperate file (assuming they don't use nested scoping). `fpdf.py` is already long enough, so I think we should take every opportunity to farm out code to other modules.

oleksii-shyman commented:
> I haven't looked at it very closely, but I expect that both `.multicell()` and `.write()` would eventually use a common internal method to actually do the wrapping.

Sounds reasonable, I think it is realistic to do it this way

oleksii-shyman commented:
The list of things I changed:

1. moved line breaking functionality into separate module. It is completely decoupled from main class. I've added _some basic cases_ to show the usage and to indicate how other unit tests will look like
2. added `_render_styled_cell_text` function (it fixed the bug with markdown rendering)
3. multi line breaking functionality should be compatible with `.write()`, because it returns chunks of _variable_ length.

I think this patch is ready for high level review. Please let me know if you have any objections or concerns, and I will add more tests and documentation.

Lucas-C commented on fpdf/line_break.py:
```diff
@@ -0,0 +1,237 @@
+            self.width += character_width
+            active_fragment.characters.append(character)
+
+    def _apply_automatic_hint(self, break_hint):
```
Could you please add a brief docstring line about what this method does?

oleksii-shyman replied:
will do

Lucas-C commented on fpdf/line_break.py:
```diff
@@ -0,0 +1,237 @@
+    def from_string(cls, string, style, underlined):
+        return cls(style, underlined, list(string))
+
+    def add_character(self, character):
```
This method is not used and can be removed

oleksii-shyman replied:
Will do

Lucas-C commented on fpdf/line_break.py:
```diff
@@ -0,0 +1,237 @@
+NEWLINE = "\n"
+
+
+class Fragment:
```
Really nice idea to structure this into a class
👍

Lucas-C commented on fpdf/line_break.py:
```diff
@@ -0,0 +1,237 @@
+        )
+
+
+text_line = namedtuple(
```
Would you mind using CamelCase for naming those `namedtuple` ?
There are classes after all, so `TextLine` / `SpaceHint` / `HyphenHint` would make more sense IMHO.

Lucas-C commented on fpdf/line_break.py:
```diff
@@ -0,0 +1,237 @@
+        original_fragment_index,
+        original_character_index,
+    ):
+        assert character != NEWLINE
```
I'm fine with defensive programming, but out of curiosity: why this check?

oleksii-shyman replied:
when user of this class tries to append `NEWLINE` character to the object of this class -> something is definitely wrong.
User is expected to create a new object of this class, and fill it with readable characters

Lucas-C replied:
Thanks for the explanation!

Lucas-C commented on fpdf/line_break.py:
```diff
@@ -0,0 +1,237 @@
+            self.fragments.append(Fragment(style, underline))
+
+        elif (
+            style != self.fragments[-1].style
```
I'm not sure to understand the role of this conditional...
Could you please add a short explanatory comment?

oleksii-shyman replied:
will do

Lucas-C commented on fpdf/line_break.py:
```diff
@@ -0,0 +1,237 @@
+        self.size_by_style = size_by_style
+        self.justify = justify
+
+        self.fragment_index = 0
```
Why are those attributes of a class instance, and not local variables inside `get_line_of_given_width()`?

oleksii-shyman replied:
`fragment_index` should definitely be an object attribute, because `get_line_of_given_width` can be called multiple times, and we need to remember how many fragments did this object already parsed.

with `size_by_style` - honestly, I didn't think too much about it. It is technically a function that is used to get the width of the character. I just didn't expect the size of the character to change in the middle of breaking the line to chunks. Is this scenario possible?

`justify` is also unlikely to change during breaking the lines to chunks.

Lucas-C replied:
I was wondering about `fragment_index` & `character_index`, and with your answer I understand they HAVE to be attributes.
You can leave `self.size_by_style` & `self.justify` as they are, they seem perfectly fine to me.

Lucas-C commented on test/test_line_break.py:
```diff
@@ -0,0 +1,95 @@
+        fragments, pdf.get_normalized_string_width_with_style
+    )
+    assert multi_line_break.get_line_of_given_width(100000) == text_line(
+        fragments, 4583, 1, False
```
Would you mind specifying constructor parameters names here, for extra readibility?

Lucas-C commented on fpdf/line_break.py:
```diff
@@ -0,0 +1,237 @@
+            )
+
+            if character == NEWLINE:
+                return current_line.manual_break()
```
When can this happen exactly?

oleksii-shyman replied:
I will add testcases that explain why this is needed.

Lucas-C reviewed (commented):
You really did an excellent job here!
The code will be much more cleaner after merging this.
The tests are also great 👍 
I made a few comments, but overall this is very good.
Could you add a mention of this change in `CHANGELOG.md` as part of this PR please?

oleksii-shyman commented:
> You really did an excellent job here! The code will be much more cleaner after merging this. The tests are also great +1 I made a few comments, but overall this is very good. Could you add a mention of this change in `CHANGELOG.md` as part of this PR please?

Thank you for your feedback, will do my changes shortly

gmischler commented:
When I submitted #291, I was prepared to wait quite a bit, as I realized it required some substantial refactoring to get it to work.
And just three months later, here we are with an evidently quite competently made implementation.
Color me impressed! 👍 

This will serve as a solid basis for someone later adapting `.write()`, so it can also align its output. Doing that will probably involve caching text chunks until each line is complete. And to make sure the last line gets processed as well, we may need some kind of "text block" context manager to finalize that. Actually, I think such text block could also manage the outline geometry, and solve #245 at the same time.
I wonder how long that would take... 😉

oleksii-shyman commented:
> When I submitted #291, I was prepared to wait quite a bit, as I realized it required some substantial refactoring to get it to work. And just three months later, here we are with an evidently quite competently made implementation. Color me impressed! +1
> 
> This will serve as a solid basis for someone later adapting `.write()`, so it can also align its output. Doing that will probably involve caching text chunks until each line is complete. And to make sure the last line gets processed as well, we may need some kind of "text block" context manager to finalize that. Actually, I think such text block could also manage the outline geometry, and solve #245 at the same time. I wonder how long that would take... wink

I have no bandwidth to handle another change of scope. Do you think that this basis is solid enough for someone later adapting `.write()` method?

Lucas-C commented:
> I have no bandwidth to handle another change of scope. Do you think that this basis is solid enough for someone later adapting .write() method?

Yes, I think it's better to get this merged first, than anyone who's willing to improve things could submit other PRs.
Please just address the review feedbacks and I'll merge this.
You will also need to rebase your branch on `master` due to this bugfix I merged earlier today: https://github.com/PyFPDF/fpdf2/pull/333

gmischler commented:
> I have no bandwidth to handle another change of scope. Do you think that this basis is solid enough for someone later adapting `.write()` method?

I didn't mean to imply that any changes to `.write()` should be a part of this PR. But yes, your clean and seperated line wrapping implementation should make them quite straightforward.

Lucas-C commented:
@oleksii-shyman please ping me if I can ever help 😊

![](https://media.giphy.com/media/xT9DPIBYf0pAviBLzO/giphy.gif)

Lucas-C commented:
This looks perfect, merging the PR now!

Lucas-C commented:
I think we should add a mention of soft-hyphen support in the docs, but I'm not sure where...
